In [ ]:
! pip install ydata_profiling

In [ ]:
!pip install hijri-converter

In [ ]:
#Import all relevant libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import zscore
from ydata_profiling import ProfileReport
from pandas_profiling import ProfileReport
from google.colab import files
from hijri_converter import convert


## This statement allows the visuals to render within your Jupyter Notebook.
%matplotlib inline

## Loading the data
We can now load the dataset into pandas using the read_csv() function. This converts the CSV file into a Pandas dataframe.

In [ ]:
#Read in the csv file and convert to a Pandas dataframe
jadarat = pd.read_csv('Jadarat_data.csv')

In [ ]:
# profile = ProfileReport(jadarat, explorative=True)  # Create profile report
# profile.to_file("report.html")  # Save as an interactive HTML file
# files.download("report.html") # Download the HTML report file  into local machine

### Viewing the dataframe
We can get a quick sense of the size of our dataset by using the shape method. This returns a tuple with the number of rows and columns in the dataset.

In [ ]:
# Display the first 3 rows
jadarat.head(3)

job_title    job_date  \
0                      محاسب  27/05/1444   
1                       بائع  27/05/1444   
2  أخصائي عمليات موارد بشرية  27/05/1444   

                                                                                                                                                                                                                                                                                                               job_desc  \
0                                                                         ['إعداد وتنظيم مستندات الصرف ومتابعة تحصيل الإيرادات وتوثيقها وتحليلها، وإعداد القيود المحاسبية وتسجيل العمليات المالية بالسجلات وتدقيق حسابات البنوك وإعداد التسويات اللازمة وإعداد التقارير والوثائق المتعلقة بالنفقات والإيرادات وحفظها.']   
1                                                                                                                                                                     ['بيع مجموعة من السلع والخدمات للعملاء، وتوفير المعلومات المتعلقة بالمنتجات واستخداماتها وأسعارها، وتولي أمور الدفع وتلقي المدفوعات من العملاء.']   
2  ['تنفيذ الإجراءات والأنظمة والنماذج الخاصة بمراقبة دوام الموظفين، والمساهمة في والإشراف على تطبيق سياسات وإجراءات الإجازات المرضية والطارئة، وتحديث المعلومات المتعلقة بالموظفين والتعامل مع شكاويهم واستفساراتهم بمهنية وإعداد الوثائق والتقارير المتخصصة والمتعلقة في مجال عمليات الموارد البشرية وعرضها وحفظها.']   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                job_tasks  \
0                                                                                                                                                                                                                                                                                                                ['   تدقيق المطالبات المالية والتأكد من اكتمال شروطها القانونية والمالية، وإعداد وتنظيم وتوثيق مستندات الصرف ومتابعة تدقيقها وإجازتها للصرف من الجهات الرقابية المعتمدة.', '  مراقبة النفقات الجارية والرأسمالية بحسب الموقف المالي الشهري لبنود الموازنة، وإعداد الموقف المالي والخلاصات الشهرية وإعداد وتنظيم الحسابات الختامية السنوية.', '  إعداد القيود المحاسبية وتسجيل العمليات المالية بالسجلات وتدقيق حسابات البنوك وإعداد التسويات اللازمة مع كشوف المطابقة الشهرية.', '  تطوير أساليب وإجراءات المحاسبة ومواكبة التطور العلمي والتقني في مجال المحاسبة.', '  إعداد الوثائق والتقارير المتخصصة في مجال محاسبة النفقات والإيرادات وحفظها في قاعدة البيانات الخاصة بذلك.', '  ']   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ['   بيع مجموعة

## 1. Data Profiling:
Data profiling is a comprehensive process of examining the data available in an existing dataset and collecting statistics and information about that data.

In [ ]:
#Display the first 3 rows
jadarat.head(3)

job_title    job_date  \
0                      محاسب  27/05/1444   
1                       بائع  27/05/1444   
2  أخصائي عمليات موارد بشرية  27/05/1444   

                                                                                                                                                                                                                                                                                                               job_desc  \
0                                                                         ['إعداد وتنظيم مستندات الصرف ومتابعة تحصيل الإيرادات وتوثيقها وتحليلها، وإعداد القيود المحاسبية وتسجيل العمليات المالية بالسجلات وتدقيق حسابات البنوك وإعداد التسويات اللازمة وإعداد التقارير والوثائق المتعلقة بالنفقات والإيرادات وحفظها.']   
1                                                                                                                                                                     ['بيع مجموعة من السلع والخدمات للعملاء، وتوفير المعلومات المتعلقة بالمنتجات واستخداماتها وأسعارها، وتولي أمور الدفع وتلقي المدفوعات من العملاء.']   
2  ['تنفيذ الإجراءات والأنظمة والنماذج الخاصة بمراقبة دوام الموظفين، والمساهمة في والإشراف على تطبيق سياسات وإجراءات الإجازات المرضية والطارئة، وتحديث المعلومات المتعلقة بالموظفين والتعامل مع شكاويهم واستفساراتهم بمهنية وإعداد الوثائق والتقارير المتخصصة والمتعلقة في مجال عمليات الموارد البشرية وعرضها وحفظها.']   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                job_tasks  \
0                                                                                                                                                                                                                                                                                                                ['   تدقيق المطالبات المالية والتأكد من اكتمال شروطها القانونية والمالية، وإعداد وتنظيم وتوثيق مستندات الصرف ومتابعة تدقيقها وإجازتها للصرف من الجهات الرقابية المعتمدة.', '  مراقبة النفقات الجارية والرأسمالية بحسب الموقف المالي الشهري لبنود الموازنة، وإعداد الموقف المالي والخلاصات الشهرية وإعداد وتنظيم الحسابات الختامية السنوية.', '  إعداد القيود المحاسبية وتسجيل العمليات المالية بالسجلات وتدقيق حسابات البنوك وإعداد التسويات اللازمة مع كشوف المطابقة الشهرية.', '  تطوير أساليب وإجراءات المحاسبة ومواكبة التطور العلمي والتقني في مجال المحاسبة.', '  إعداد الوثائق والتقارير المتخصصة في مجال محاسبة النفقات والإيرادات وحفظها في قاعدة البيانات الخاصة بذلك.', '  ']   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ['   بيع مجموعة

In [ ]:
# Print the number of rows and columns in the dataset
jadarat.shape

(1470, 18)

In [ ]:
# Print the label of the columns
jadarat.columns

Index(['job_title', 'job_date', 'job_desc', 'job_tasks', 'comp_name',
       'comp_no', 'comp_type', 'comp_size', 'eco_activity', 'qualif', 'region',
       'city', 'benefits', 'contract', 'positions', 'job_post_id', 'exper',
       'gender'],
      dtype='object')

**Columns Explanation:**

In [ ]:
# Improve the display of long text within the DataFrame
pd.set_option('display.max_colwidth', None)
df_info = pd.read_csv('Jadarat_Data_Description.csv') # Load the dataset from the CSV file into a DataFrame
df_info

Column Name          Symbol / Code  \
0      job_title       Simplified Roles   
1       job_date      Dates (Gregorian)   
2      comp_name                  AS-IS   
3      comp_type                   1, 0   
4      comp_size       MA, MB, MC, etc.   
5   eco_activity        Various Sectors   
6         region  Names (e.g., Eastern)   
7           city       Translated Names   
8       contract                   1, 0   
9       benefits                   1, 0   
10     positions     Numbers (e.g., 10)   
11         exper         Numeric values   
12        gender                0, 1, 2   
13        Salary         Numeric values   

                                                                      Explanation  
0   Core job roles mapped to predefined categories (e.g., Accountant, Technician)  
1                                                         Date the job was posted  
2                  The company name remains as originally provided in the dataset  
3                                              1 - Private, 0 - Semi-Governmental  
4                   Indicates the company size: MA - Medium A, SB - Small B, etc.  
5                Generalized industry sector such as Trade and Retail, Healthcare  
6                Translated and normalized Saudi regions (e.g., Eastern Province)  
7                  Translated and normalized city names (e.g., AD DAMMAM, RIYADH)  
8                                                       1 - Full-time, 0 - Remote  
9                                           1 - Benefits offered, 0 - No benefits  
10                                      Number of positions available for the job  
11               Years of experience required, converted to integers (e.g., 0, 2)  
12                                                 0 - Male, 1 - Female, 2 - Both  
13                                                   Represents the salary in SAR

In [ ]:
# Print the inforamtion about the data like missing values and data types
jadarat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   job_title     1470 non-null   object
 1   job_date      1470 non-null   object
 2   job_desc      1470 non-null   object
 3   job_tasks     1470 non-null   object
 4   comp_name     1470 non-null   object
 5   comp_no       1470 non-null   object
 6   comp_type     1470 non-null   object
 7   comp_size     1456 non-null   object
 8   eco_activity  1380 non-null   object
 9   qualif        780 non-null    object
 10  region        1470 non-null   object
 11  city          1470 non-null   object
 12  benefits      1470 non-null   object
 13  contract      1470 non-null   object
 14  positions     1470 non-null   object
 15  job_post_id   1470 non-null   int64 
 16  exper         1470 non-null   object
 17  gender        1470 non-null   object
dtypes: int64(1), object(17)
memory usage: 206.8+ KB


**Missing Values =	794**

The dataset have missing values in three columns:

- `comp_size` has 14 missing values.
- `eco_activity` has 90 missing values.
- `qualif` has 690 missing values.


**Data Types:**

All columns have the correct data types EXCEPT for `comp_no`, `positions` and `exper` need to be intger. The `job_date` should be converted to a datetime type.

**Note1:**We prefer converting Hijri dates to Gregorian for the following reasons:

**1- Compatibility with Analytical Tools**

Most data analysis tools (such as pandas and matplotlib) rely on the Gregorian calendar when handling time series data.

**2- Easier Comparisons with Other Datasets**

The majority of economic and financial data use the Gregorian calendar, making it easier to merge labor market data with economic datasets.

**3- More Accurate Statistical Analyses**

Time Series Analysis can be performed without dealing with complex conversions, improving the accuracy and efficiency of the analysis.

**Note2:**
Any column that can be represented as binary values will be converted to facilitate further analysis. Like: `comp_type`, `contract`, `gender`.

In [ ]:
jadarat.describe(include='object')

job_title    job_date  \
count       1470        1470   
unique       324          92   
top         بائع  09/06/1444   
freq          98          47   

                                                                                                                                                 job_desc  \
count                                                                                                                                                1470   
unique                                                                                                                                                438   
top     ['بيع مجموعة من السلع والخدمات للعملاء، وتوفير المعلومات المتعلقة بالمنتجات واستخداماتها وأسعارها، وتولي أمور الدفع وتلقي المدفوعات من العملاء.']   
freq                                                                                                                                                   93   

                                                                                                                                                                                                                                                                                                                                                            job_tasks  \
count                                                                                                                                                                                                                                                                                                                                                            1470   
unique                                                                                                                                                                                                                                                                                                                                                            444   
top     ['   بيع مجموعة من السلع والخدمات للعملاء.', '  توفير المعلومات المتعلقة بالمنتجات، والاستخدامات، والأسعار.', '  مراقبة مستويات المخزون والحفاظ على ترتيب المتجر، وإصدار طلبات لشراء المنتجات، والتعامل مع المنتجات الواردة.', '  تولي أمور الدفع وتلقي المدفوعات من العملاء، ومعالجة طلبات استبدال المنتجات، والمرتجعات وفقاً للسياسة المتبعة.', '  ', '  ']   
freq                                                                                                                                                                                                                                                                                                                                                               90   

                                                          comp_name   comp_no  \
count                                                          1470      1470   
unique                                                          886       894   
top     شركه العرض المتقن للخدمات التجارية شركة مساهمة سعودية مقفلة  1-169123   
freq                                                             16        16   

       comp_type    comp_size       eco_activity  \
count       1470         1456               1380   
unique         2            7                361   
top          خاص  صغيرة فئة ب  المطاعم مع الخدمة   
freq        1465          635                 65   

                                          qualif  region          city  \
count                                        780    1470          1470   
unique                                       615      14           114   
top     ['Language data', 'الانجليزية', 'متوسط']  الرياض  AR RIYADH...   
freq                                          28     622           523   

                    benefits   contract positions    exper gender  
count                   1470       1470      1470     1470   1470  
unique                   393          2        25        6      3  
top     ['Sa

In [ ]:
jadarat['city'].value_counts().head(50)

city
AR RIYADH...    523
JEDDAH...       245
AD DAMMAM...     78
MAKKAH AL...     72
AL KHUBAR...     54
AL MADINA...     51
ABA ASLAB...     51
AL BAHA...       24
HAIL...          17
AL KHARJ...      16
KHAMIS MU...     14
HAFAR AL ...     14
AL QUNFID...     13
YANBU AS ...     12
BURAYDAH...      12
AL JUBAIL...     12
AT TAIF...       12
AL HUFUF...      11
ABHA...          10
DHAHRAN...       10
AL MUBARR...      9
UNAYZAH...        9
ABRAQIYAH...      9
AL QATIF...       8
MALHAM...         8
ABA AL QI...      8
ABA AL BA...      7
YANBU...          6
GHITHAH...        6
NAJRAN...         6
ABRAQ AN ...      5
AL MUKHWA...      5
ATH THUQB...      4
QILWAH...         4
AL AQIQ...        4
YANBU AL ...      4
AL BUKAYR...      3
NUMARAH...        3
SAKAKA...         3
AL HALQAH...      3
BILJURASH...      3
AL HAWIYA...      3
AL BADAI'...      3
TANUMAH...        3
AS SALAMA...      3
ABA AL KI...      2
AL MUZAHI...      2
AL HAIR...        2
TABUK...          2
AL 'AZIZI...      2
Name: count, dtype: int64

In [ ]:
jadarat[jadarat['city'] == '%ABA']

Empty DataFrame
Columns: [job_title, job_date, job_desc, job_tasks, comp_name, comp_no, comp_type, comp_size, eco_activity, qualif, region, city, benefits, contract, positions, job_post_id, exper, gender]
Index: []

Insights:
- The high demand for sales positions suggests that the retail sector is a significant employer in the job market.
- With 324 unique job roles, there is a diverse range of employment opportunities, indicating a dynamic job market.
- A large number of unique companies (886) suggests a diverse and competitive job market.
- A single company `شركة العرض المتقن للخدمات التجارية شركة مساهمة سعودية مقفلة` posting 16 job listings may indicate:
  - The company is actively expanding.
  - Has high turnover rate requiring frequent hiring.
- The private sector dominates the job market in this dataset, with the majority of job postings being from private companies.
- Most job postings are concentrated in major cities like Riyadh.
- The dataset contains diverse industries (361 different activities), showing that job opportunities span multiple sectors.
- Economic activities are diverse, but some industries dominate, such as food services.
- Some regions may have significantly more job opportunities than others (Total Regions = 14).
- The dataset includes a wide range of cities (114), indicating nationwide job availability.
-  Many jobs may not require experience, making them more accessible for fresh graduates.



In [ ]:
jadarat.describe()

job_post_id
count  1.470000e+03
mean   2.020203e+13
std    5.863941e+04
min    2.020203e+13
25%    2.020203e+13
50%    2.020203e+13
75%    2.020203e+13
max    2.020203e+13

The process of profiling differs slightly for categorical and numerical variables due to their inherent differences.

**The two main types of data are:**
- Quantitative (numerical) data
- Qualitative (categorical) data

### Data Quality Checks
Data quality checks involve the process of ensuring that the data is accurate, complete, consistent, relevant, and reliable.


**Here are typical steps involved in checking data quality:**

#### 1. Reliability:
Evaluate the data's source and collection process to determine its trustworthiness.

The dataset is from (**Jadarat**) website and it is a unified national platform for employment in Saudi Arabia. The platform is supervised by the **Saudi Ministry of Human Resources and Social Development**, which enhances its reliability and validity.

#### 2. Timeliness:
Ensure the data is up-to-date and reflective of the current situation or the period of interest for the analysis.

Since the questions listed in the problem statement are not identifying specific period of time or if they need it over the years, so the dataset need to be up-to-date and the most current dataset is this.

#### 3. Consistency:

Confirm that the data is consistent within the dataset and across multiple data sources. For example, the same data point should not have different values in different places.


The dataset is consistent because we do not have multiple sources that can change the meaning of a value. So, the value is remains logically valid.

#### 4. Relevance:
Assess whether the data is appropriate and applicable for the intended analysis. Data that is not relevant can skew results and lead to incorrect conclusions.

**Key considerations for relevance include:**

> 1. Sample Appropriateness: Confirm that your data sample aligns with your analysis objectives. For instance, utilizing data from the Northern region will not yield accurate insights for the Western region of the Kingdom.
>
> 2. Variable Selection: Any column will not be relevant for our analysis, we can get rid of these using the drop() method. We will set the “axis” argument to 1 since we’re dealing with columns, and set the “inplace” argument to True to make the change permanent.


**1- Sample Appropriateness:** This dataset are relevance to the problem statement question.

**2- Variable Selection:** All columns are relevant for our analysis EXCEPT 3 columns `job_desc`, `job_tasks`, `qualif` and `comp_no	` we will drop them.



In [ ]:
# go to drop unrelevant columns

#### 5. Uniqueness:
Check for and remove duplicate records to prevent skewed analysis results.


In [ ]:
jadarat.duplicated().sum()

11

In [ ]:
jadarat[jadarat.duplicated(keep=False)]

job_title    job_date  \
120                           نادل  24/04/1444   
139                           نادل  24/04/1444   
871                     كاتب اتصال  05/06/1444   
954                           بائع  07/06/1444   
1015  فني كهربائي تمديدات كهربائية  04/06/1444   
1120                         محاسب  09/06/1444   
1160  فني كهربائي تمديدات كهربائية  04/06/1444   
1164                      كاتب شحن  02/06/1444   
1207            كاتب علاقات حكومية  16/06/1444   
1258            كاتب علاقات حكومية  16/06/1444   
1263             أخصائي علاج طبيعي  14/06/1444   
1310                         محاسب  23/06/1444   
1313             أخصائي خدمة عملاء  23/06/1444   
1371             أخصائي علاج طبيعي  14/06/1444   
1381                         محاسب  09/06/1444   
1382                          بائع  07/06/1444   
1385                    كاتب اتصال  05/06/1444   
1389                      كاتب شحن  02/06/1444   
1457                         محاسب  23/06/1444   
1458             أخصائي خدمة عملاء  23/06/1444   
1467                         صيدلي  18/06/1444   
1469                         صيدلي  18/06/1444   

                                 comp_name comp_type     comp_size  \
120       مطعم ملعقة الطاهي لتقديم الوجبات       خاص   صغيرة فئة ب   
139       مطعم ملعقة الطاهي لتقديم الوجبات       خاص   صغيرة فئة ب   
871             شركة اساس الخيرات الزراعية       خاص   صغيرة فئة ب   
954                   مؤسسة شذرات للسراميك       خاص   صغيرة فئة ب   
1015      شركة صقور العرب للمقاولات العامة       خاص   صغيرة فئة ب   
1120          شركة اتقان الصيانة للمقاولات       خاص  متوسطة فئة ب   
1160      شركة صقور العرب للمقاولات العامة       خاص   صغيرة فئة ب   
1164  شركة الناقل العاجل للخدمات اللوجستية       خاص   صغيرة فئة ب   
1207  مجموعة اجير المتحدة للتشغيل والصيانة       خاص   صغيرة فئة أ   
1258  مجموعة اجير المتحدة للتشغيل والصيانة       خاص   صغيرة فئة أ   
1263                   مركز الراحة للحجامة       خاص   صغيرة فئة ب   
1310                    شركة بوركت للتجارة       خاص   صغيرة فئة ب   
1313                    شركة بوركت للتجارة       خاص   صغيرة فئة ب   
1371                   مركز الراحة للحجامة       خاص   صغيرة فئة ب   
1381          شركة اتقان الصيانة للمقاولات       خاص  متوسطة فئة ب   
1382                  مؤسسة شذرات للسراميك       خاص   صغيرة فئة ب   
1385            شركة اساس الخيرات الزراعية       خاص   صغيرة فئة ب   
1389  شركة الناقل العاجل للخدمات اللوجستية       خاص   صغيرة فئة ب   
1457                    شركة بوركت للتجارة       خاص   صغيرة فئة ب   
1458                    شركة بوركت للتجارة       خاص   صغيرة فئة ب   
1467   شركة تقنية الدواجن والألبان للتجارة       خاص   صغيرة فئة ب   
1469   شركة تقنية الدواجن والألبان للتجارة       خاص   صغيرة فئة ب   

                                                                                                                       eco_activity  \
120   تشغيل المقاصف والكفتيريات على أساس إمتياز محدد ، يشمل ( المقاصف والكفتيريات في المصانع والمكاتب والمستشفيات والمدارس .. الخ )   
139   تشغيل المقاصف والكفتيريات على أساس إمتياز محدد ، يشمل ( المقاصف والكفتيريات في المصانع والمكاتب والمستشفيات والمدارس .. الخ )   
871                                                                                           البيع بالجملة للفواكة والثمار الزيتية   
954                                                           البيع بالتجزئة للرخام والحجر الطبيعي والاصطناعي والسيراميك والبورسلان   
1015                                                                                      قصور وصالات الافراح والمناسبات مع الاقامة   
1120                                                                                                                            NaN   
1160                                                                                      قصور وصالات الافراح والمناسبات مع الاقامة   
1164                                                                                 المخازن العامة التي تضم مجموعة متنوعة من السلع   
1207                                            

In [ ]:
# go to delete duplicates columns

#### 6. Completeness:
Ensure that no critical data is missing. This might mean checking for null values or required fields that are empty.

We will start by checking the dataset for missing or null values. For this, we can use the isna() method which returns a dataframe of boolean values indicating if a field is null or not. To group all missing values by column, we can include the sum() method.

In [ ]:
#Display number missing values per column

> After convert the date to gregorian, 47 missing values are appeared in `job_date_gregorian` column.

In [ ]:
# go to clean them

#### 7. Check Accuracy:

Verify that the data is correct and precise. This could involve comparing data samples with known sources or using validation rules.

**The process includes:**
1. Validating the appropriateness of data types for the dataset.
2. Identifying outliers  using established validation  rule

In [ ]:
# check columns types

There are 3 columns need to be convered into correct data types:

1- `job_date`: We prefer to keep this column and create a new one based on it to be date in gregorian. After that, convert it to datetime type.

2- `exper`: Extract numeric values from this column and then convert it to integer.

3- `positions`: Extract only numeric values from this column and then convert it to integer.


In [ ]:
# go to clean them

There are columns it is good to convert it to Binary to make analysis:



In [ ]:
jadarat['contract'].value_counts()

contract
دوام كامل     1437
عمل عن بعد      22
Name: count, dtype: int64

Current values: "Full-time" and "Remote"

After conversion:

1 = Full-time

0 = Remote

In [ ]:
# go to clean them

In [ ]:
# check outliers

**What is an Outlier?**
Outlier is an row/observation that appears far away and diverges from an overall pattern in a sample.

**What are the types of Outliers?**
1. Univariate: These outliers can be found when we look at distribution of a single variable
2. Multivariate: are outliers in an n-dimensional space. In order to find them, you have to look at distributions in multi-dimensions. example (hight=100, weight=100) for a person

**What causes Outliers?**
Whenever we come across outliers, the ideal way to tackle them is to find out the reason of having these outliers. The method to deal with them would then depend on the reason of their occurrence.

Let’s understand various types of outliers:

1. Data Entry Errors:- Human errors such as errors caused during data collection, recording, or entry can cause outliers in data.
2. Measurement Error: It is the most common source of outliers. This is caused when the measurement instrument used turns out to be faulty.
3. Data Processing Error: Whenever we perform data mining, we extract data from multiple sources. It is possible that some manipulation or extraction errors may lead to outliers in the dataset.
4. Sampling error: For instance, we have to measure the height of athletes. By mistake, we include a few basketball players in the sample. This inclusion is likely to cause outliers in the dataset.
5. Natural Outlier: When an outlier is not artificial (due to error), it is a natural outlier. For instance: In my last assignment with one of the renowned insurance company, I noticed that the performance of top 50 financial advisors was far higher than rest of the population. Surprisingly, it was not due to any error. Hence, whenever we perform any data mining activity with advisors, we used to treat this segment separately.


**What is the impact of Outliers on a dataset?**


![image.png](https://www.analyticsvidhya.com/wp-content/uploads/2015/02/Outlier_31.png)



**How to detect Outliers?**

1. Most commonly used method to detect outliers is visualization (Univariate Graphical Analysis).

We use 3 common visualization methods:
>- Box-plot: A box plot is a method for graphically depicting groups of numerical data through their quartiles. The box extends from the Q1 to Q3 quartile values of the data, with a line at the median (Q2). The whiskers extend from the edges of the box to show the range of the data. Outlier points are those past the end of the whiskers. Box plots show robust measures of location and spread as well as providing information about symmetry and outliers.
>
>  
>![image.png](https://miro.medium.com/v2/resize:fit:698/format:webp/1*VK5iHA2AB28HSZwWwUbNYg.png)
>
>
>- Histogram
>- Scatter Plot: A scatter plot is a mathematical diagram using Cartesian coordinates to display values for two variables for a set of data. The data are displayed as a collection of points, each having the value of one variable determining the position on the horizontal axis and the value of the other variable determining the position on the vertical axis. The points that are far from the population can be termed as an outlier.
>
>  
>![image.png](https://miro.medium.com/v2/resize:fit:4800/format:webp/1*Ov6aH-8yIwNoUxtMFwgx4g.png)
>
>

2. Using statistical method (Univariate Non-Graphical analysis):
>- Any value, which is beyond the range of -1.5 x IQR to 1.5 x IQR

![image.png](https://www.whatissixsigma.net/wp-content/uploads/2015/07/Box-Plot-Diagram-to-identify-Outliers-figure-1.png)

>- Use capping methods. Any value which out of range of 5th and 95th percentile can be considered as outlier
>- Data points, three or more standard deviation away from mean are considered outlier: The Z-score is the signed number of standard deviations by which the value of an observation or data point is above the mean value of what is being observed or measured. While calculating the Z-score we re-scale and center the data and look for data points that are too far from zero. These data points which are way too far from zero will be treated as the outliers. In most of the cases, a threshold of 3 or -3 is used i.e if the Z-score value is greater than or less than 3 or -3 respectively, that data point will be identified as outliers.
> - Outlier detection is merely a special case of the examination of data for influential data points and it also depends on the business understanding


In [ ]:
# go to univariate graphical analysis
# go to lesson : data visualisation 1 - chart type section
# then go to univariate graphical analysis
# detect outliers using graphs varbaly

In [ ]:
# go to lesson: statistics 1 then statistics 3
# then go to univariate Non graphical analysis
# detect outliers using numerical statistics

In [ ]:
# go to delete ouliers

## 2. Data Cleaning:

Preliminary findings from data profiling can lead to cleaning the data by:
- Handling missing values
- Correcting errors.
- Dealing with outliers.

-------------------



### Handling missing values:

**Why my data has missing values?**
They may occur at two stages:
1. Data Extraction: It is possible that there are problems with extraction process. Errors at data extraction stage are typically easy to find and can be corrected easily as well.
2. Data collection: These errors occur at time of data collection and are harder to correct.

**Why do we need to handle the missing data?**
To avoid:
- Bias the conclusions.
- Leading the business to make wrong decisions.

**Which are the methods to treat missing values ?**
1. Deletion: we delete rows where any of the variable is missing. Simplicity is one of the major advantage of this method, but this method reduces the power of model because it reduces the sample size.

2. Imputation: is a method to fill in the missing values with estimated ones. This imputation is one of the most frequently used methods.

    2.1. Mean/ Mode/ Median Imputation: It consists of replacing the missing data for a given attribute by the mean or median (quantitative attribute) or mode (qualitative attribute) of all known values of that variable.
    > It can be of two types:
    > - Generalized Imputation: In this case, we calculate the mean or median for all non missing values of that variable then replace missing value with mean or median.
    > - Similar case Imputation: In this case, we calculate average for each group individually of non missing values then replace the missing value based on the group.

    2.2. Constant Value
   
    2.3. Forward Filling
   
    2.4. Backward Filling

6. Prediction Model:  Prediction model is one of the sophisticated method for handling missing data. Here, we create a predictive model to estimate values that will substitute the missing data.  In this case, we divide our data set into two sets: One set with no missing values for the variable and another one with missing values. First data set become training data set of the model while second data set with missing values is test data set and variable with missing values is treated as target variable. Next, we create a model to predict target variable based on other attributes of the training data set and populate missing values of test data set.

> There are 2 drawbacks for this approach:
> - The model estimated values are usually more well-behaved than the true values
> - If there are no relationships with attributes in the data set and the attribute with missing values, then the model will not be precise for estimating missing values.

9. KNN Imputation: In this method of imputation, the missing values of an attribute are imputed using the given number of attributes that are most similar to the attribute whose values are missing. The similarity of two attributes is determined using a distance function. It is also known to have certain advantage & disadvantages.

   > **Advantages:**
   > - k-nearest neighbour can predict both qualitative & quantitative attributes
   > - Creation of predictive model for each attribute with missing data is not required
   > - Attributes with multiple missing values can be easily treated
   > - Correlation structure of the data is taken into consideration

   > **Disadvantage:**
   > - KNN algorithm is very time-consuming in analyzing large database. It searches through all the dataset looking for the most similar instances.
   > - Choice of k-value is very critical. Higher value of k would include attributes which are significantly different from what we need whereas lower value of k implies missing out of significant attributes.

--------------------


In [ ]:
jadarat.drop(columns=['job_desc', 'job_tasks', 'comp_no', 'qualif'], inplace=True)

In [ ]:
# go back to 5th dimention --> Uniqueness

In [ ]:
jadarat.drop_duplicates(keep='first', inplace=True)
print(jadarat.duplicated().sum())  # Should return 0 if all duplicates are removed

0


In [ ]:
# go back to 6th dimention --> Completeness

In [ ]:
# go back to 7th dimension Accuracy

In [ ]:
# Convert the date from hijri to gregorian
# Import hijri_converter module from convert library
from hijri_converter import convert


def hijri_to_gregorian(hijri_date):
    try:
        day, month, year = map(int, hijri_date.split('/'))

        if not (1 <= month <= 12 and 1300 <= year <= 1500):
            return None

        return convert.Hijri(year, month, day).to_gregorian()

    except Exception as e:
        return None

In [ ]:
jadarat['job_date_gregorian'] = jadarat['job_date'].apply(hijri_to_gregorian)
jadarat[['job_date', 'job_date_gregorian']].head()

job_date job_date_gregorian
0  27/05/1444         2022-12-21
1  27/05/1444         2022-12-21
2  27/05/1444         2022-12-21
3  27/05/1444         2022-12-21
4  27/05/1444         2022-12-21

In [ ]:
jadarat.isnull().sum()

job_title              0
job_date               0
comp_name              0
comp_type              0
comp_size             14
eco_activity          90
region                 0
city                   0
benefits               0
contract               0
positions              0
job_post_id            0
exper                  0
gender                 0
job_date_gregorian    47
dtype: int64

Missing values in job_date_gregorian

In [ ]:
print(jadarat[['job_date_gregorian']].dtypes)

job_date_gregorian    object
dtype: object


In [ ]:
jadarat['job_date_gregorian'] = pd.to_datetime(jadarat['job_date_gregorian'], errors='coerce')
print(jadarat[['job_date_gregorian']].dtypes)

job_date_gregorian    datetime64[ns]
dtype: object


In [ ]:
jadarat['exper'] = jadarat['exper'].astype(str).str.extract(r'(\d+)')
# Extract only the second number from 'positions' (after '/')
jadarat['positions'] = jadarat['positions'].astype(str).str.extract(r'/\s*(\d+)')

In [ ]:
# Convert `positions` and `exper` to integer
columns_to_int = ['positions', 'exper']
for col in columns_to_int:
    jadarat[col] = pd.to_numeric(jadarat[col], errors='coerce').astype('Int64')

print(jadarat[['positions']].dtypes)
print(jadarat[['exper']].dtypes)

positions    Int64
dtype: object
exper    Int64
dtype: object


In [ ]:
jadarat.describe()

positions   job_post_id     exper             job_date_gregorian
count     1470.0  1.470000e+03    1470.0                           1423
mean    2.938776  2.020203e+13  1.278912  2022-12-09 21:43:23.232607232
min          1.0  2.020203e+13       0.0            2022-10-01 00:00:00
25%          1.0  2.020203e+13       0.0            2022-11-16 00:00:00
50%          1.0  2.020203e+13       0.0            2022-12-15 00:00:00
75%          2.0  2.020203e+13       2.0            2023-01-01 00:00:00
max         50.0  2.020203e+13      12.0            2023-01-23 00:00:00
std     6.400864  5.863941e+04  1.879568                            NaN

### Correcting errors

-------------------

### Dealing with outliers:

**How to remove Outliers?**
Most of the ways to deal with outliers are similar to the methods of missing values like deleting rows, transforming them, binning them, treat them as a separate group, imputing values and other statistical methods. Here, we will discuss the common techniques used to deal with outliers:

1. Deleting rows: We delete outlier values if it is due to data entry error, data processing error or outlier rows are very small in numbers. We can also use trimming at both ends to remove outliers.

2. Imputing: Like imputation of missing values, we can also impute outliers. We can use mean, median, mode imputation methods. Before imputing values, we should analyse if it is natural outlier or artificial. If it is artificial, we can go with imputing values. We can also use statistical model to predict values of outlier rows and after that we can impute it with predicted values.

3. Treat separately: If there are significant number of outliers, we should treat them separately in the statistical model. One of the approach is to treat both groups as two different groups and build individual model for both groups and then combine the output.


## 3. Univariate Analysis:

This involves examining single variables to understand their characteristics (distribution, central tendency, dispersion, and shape).

We calculate **numerical values** about the data that tells us about the distribution of the data. We also **draw graphs** showing visually how the data is distributed. **To answer the following questions about Features/characteristics of Data:**
- Where is the center of the data? (location)
- How much does the data vary? (scale)
- What is the shape of the data? (shape)

**The benefits of this analysis:**
Statistics summary gives a high-level idea to identify whether the data has any outliers, data entry error, distribution of data such as the data is normally distributed or left/right skewed

**In this step, we will explore variables one by one using following approaches:**

### 1. Univariate Graphical Analysis:
Method to perform uni-variate analysis will depend on whether the variable type is categorical or numerical.

#### I. Categorical Variables:

we’ll use frequency table to understand distribution of each category
- Bar Chart (Ordinal) - Orderd
- Pie Chart (Nominal) - non Orderd

#### II. Numerical Variables:

we need to understand the central tendency and spread of the variable (Descriptive Analysis) using:
   - Box plot
   - Histogram

### 2. Univariate Non-Graphical analysis:

- Where is the center of the data? (location) --> **Measures of central tendency**
- How much does the data vary? (scale) --> **Measure of variability**
- What is the shape of the data? (shape) --> **Measures of variation combined with an average (measure of center) gives a good picture of the distribution of the data.**

## 4. Bivariate/Multivariate Analysis:

Here, you look at the relationships between two or more variables. This can involve looking for correlations, patterns, and trends that suggest a relationship or an association.

We can perform bi-variate analysis for any combination of categorical and numerical variables. The combination can be:
| bi-variate variables   | Plot type |
| ------------- | ------------- |
| Categorical & Categorical| Stacked Bar Chart |
| Categorical & numerical  | scatter plot, histogram, box plot|
| numerical  & numerical  | Scatter plot, line chart|


Multivariate Analysis:
- Heat map
- Bar Chart
- Scatter Chart
- Line Chart

**Categorical & Categorical --> (Stacked Column Chart)**

**Categorical & numerical --> (scatter plot, histogram, box plot)**

**numerical & numerical --> (Scatter plot, line chart)**

We could also use a correlation matrix to get more specific information about the relationship between these two variables.